kernel/src/main/scala/me/rexim/morganey/syntax/LambdaParser.scala

```diff
@@ -27,38 +28,65 @@ class LambdaParser extends JavaTokenParsers with ImplicitConversions {
   def variable: Parser[LambdaVar] =
     identifier.r ^^ { LambdaVar }
 
-  def numericLiteral: Parser[LambdaTerm] =
+  def validNumberLiteral: Parser[Int] =
     numberLiteral.r ^? ({
-      case IntMatcher(x) => ChurchNumberConverter.encodeNumber(x)
+      case IntMatcher(x) => x
     }, { (rawInt) => s"`$rawInt' is too big"})
 
-  def characterLiteral: Parser[LambdaTerm] = (
+  def numericLiteral: Parser[LambdaTerm] =
+    validNumberLiteral ^^ ChurchNumberConverter.encodeNumber
+
+  def validCharacterLiteral: Parser[Char] = (
     escapedCharLiteral.r ^^ { s =>
-      ChurchNumberConverter.encodeNumber(escapeSequences(s charAt 2))
-    }
-    | symbolCharLiteral.r ^^ { s =>
-      ChurchNumberConverter.encodeNumber(s charAt 1)
+      escapeSequences(s charAt 2)
     }
+    | symbolCharLiteral.r ^^ { _ charAt 1 }
   )
 
+  def characterLiteral: Parser[LambdaTerm] =
+    validCharacterLiteral ^^ { c => ChurchNumberConverter.encodeNumber(c.toInt) }
+
   def stringLiteralTerm: Parser[LambdaTerm] =
     stringLiteral ^^ { s =>
       ChurchPairConverter.encodeString(unquoteString(s))
     }
 
+  private def range[T](p: => Parser[T])(implicit ev: Integral[T]): Parser[LambdaTerm] = {
+    val q = p
+    val parser = q ~ opt(comma ~> q) ~ (rangeOperator ~> q)
+    parser ^^ { case start ~ next ~ exit =>
+      val step  = next.map(ev.minus(_, start)).getOrElse(ev.one)
+      val range = NumericRange.inclusive(start, exit, step).toList
+      val nums  = range map ev.toInt map ChurchNumberConverter.encodeNumber
+      ChurchPairConverter.encodeList(nums)
+    }
+  }
```

Comment:
Morganey is an implementation of Untyped Lambda Calculus and there are no types indeed in it. Everything is a Lambda Term and how it's rendered is just a matter of perspective.

Some interesting examples on that:

1.

```
λ> 97
char: a
λ> 'a'
char: a
```

2.

```
λ> []
number: 0
```

3.

```
λ> (\f . I)
number: 0
```

This happens because the REPL tries to detect things in the follow order:
- char
- number
- list of chars
- list of numbers
- term

But we could detect them in a different order and the result would be different. But the point is in Morganey everything is a term no matter how it's presented.

In that case, at least for me as a Morganey programmer, it is pretty intuitive to expect that the following thing would work:

```
λ> [1 .. 'a']
[1.7] failure: string matching regex `[0-9]+' expected but `'' found

[1 .. 'a']
      ^
```

`1` and `'a'` are not literals of different types. There are no types in Morganey. Morganey is an implementation of Untyped Lambda Calculus. Those are just different ways to present the same thing - Lambda Terms. So, I'd expect `[1 .. 'a']` to be equivalent to `[1 .. 97]`.

Also, the bounds of ranges cannot be expressions for some reason!

```
λ> [(1 1) .. (2 2)]
[1.8] failure: `]' expected but `.' found

[(1 1) .. (2 2)]
       ^
λ>
```

While in Haskell it's totally possible

```
GHCi, version 7.10.3: http://www.haskell.org/ghc/  :? for help
Prelude> [1 + 1 .. 2 ^ 3]
[2,3,4,5,6,7,8]
Prelude>
```

Maybe the ranges are not so literal after all. ;)

They are clearly expressions. And what we see here is not a literal, but an expression that for some unknown for me reason is evaluated outside of the language right in the parser. It feels so out of this world. It's not a Morganey expression. It's a Haskell expression (a cutted one, though, since it's not possible put an expression as a boundary) slaped onto Morganey language. And, I'm really sorry, I don't like that at all.

---

Ok, let's try to think how the Morganey ranges (not the Haskell ones) should be implemented.

First, I think we should allow any term as the boundary. Second, evaluate range inside of Morganey rather then inside of the Parser. So, let's just transform `[x .. y]` into `[x, (SUCC x), (SUCC (SUCC x)), ..., y]` and let the reduction do all the work with evaluating all of this `SUCC`-s! `x` is not a number?  Who cares! There no types in Morganey! You put it there, you asked for this exact expression: `[x, (SUCC x), (SUCC (SUCC x)), ..., y]`. It's gonna be evaluating into something. Probably a garbage. That's the Morganey way. And it always was. Everything is a term. Syntax is just sugar to **construct** terms more quickly, not to **evaluate** them.

But the above is just a concept. It doesn't answer many questions. Such as, "How to transform `[x, y .. z]`?", "When to stop wrapping `SUCC`-s over the `x`?", "How to know what to use `SUCC` or `PREV`?" and all of that should not involve evaluating terms during the parsing.

In that case, I'd probably add the following function to the prelude:

```morganey

range := (\left . (\step . (\right . <generate the range>)))

```

and would transform `[x0, x1 .. y]` into `(((range x0) ((minus x1) x0)) y)`. And something like that I'd call a true Morganey range. But for that we need to finally implement the Prelude mechanism #79, so the `range` is always present in any Morganey session. Also resolve the dependent bindings problems #109, since the modules are almost unusable because of it. And that is why I said that the ranges should be a separate issue, because I predicted to see something like this. I simply didn't want to say anything until there is code. Just don't like to discuss unexisting things.

---

I'm gonna merge this only under the following conditions:

- We create a separate issue for implementing ranges the Morganey way and removing all of that from the parsing process.
- We give that issue the highest priority (I'm gonna work on it and its blockers on Morganey Tsoding sessions),
- Anything like that in the future will be immediately declined.

Deal?